sdk/clientcore/tools/annotation-processor/src/main/java/io/clientcore/tools/codegen/utils/PathBuilder.java

```diff
@@ -0,0 +1,86 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package io.clientcore.tools.codegen.utils;
+
+import io.clientcore.tools.codegen.exceptions.MissingSubstitutionException;
+import io.clientcore.tools.codegen.models.HttpRequestContext;
+import io.clientcore.tools.codegen.models.Substitution;
+
+import java.util.Map;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class PathBuilder {
+    // this class takes a 'raw host' string that contains {} delimited parameters, and needs to convert it into a
+    // Java string concatenation that can be used in the generated code. For example, the raw host string:
+    // https://{endpoint}/keys/{path1}
+    // would be converted into:
+    // "https://" + endpointParam + "/keys/" + pathValue
+    // Note that query parameters may also exist, and should be appended to the end of the URL string using
+    // a Map containing key-value pairs.
+    // Note that the 'endpoint' parameter is special - it is always the first parameter, and is always a host parameter.
+    public static String buildPath(String rawHost, HttpRequestContext method) {
+        if (method == null) {
+            throw new NullPointerException("method cannot be null");
+        }
+
+        final boolean hasQueryParams = !method.getQueryParams().isEmpty();
+
+        Pattern pattern = Pattern.compile("\\{(.+?)\\}");
```

Comment:
nit: Make this a constant instead of creating a new Pattern each time as it never changes.